### PR TITLE
Make `DynamicType` a standalone library

### PR DESCRIPTION
**Pull request by zasdfgbnm**
This PR split out `DynamicType` from nvFuser to maintain it as a separate C++ template library. The codebase of this library is located in `Fuser/lib/dynamic_type`. The main reason for doing so is because I want to be able to build the test for `DynamicType` separately without building the entire nvFuser. This is important for the productivity of the development of `DynamicType`, because I need to do a lot of try-and-errors in development, and each change of the `dynamic_type.h` requires rebuilding almost the entire nvFuser, which today is very slow (due to `DynamicType`). I want to be able to ignore the rest of nvFuser and just try and error in the test of `test_dynamic_type.cpp`. In the future, `test_dynamic_type.cpp` will be split into multiple separate files for better parallelizing the build. By pulling `DynamicType` as a separate library, I think we can get better code organization, because, for example, we can put the documentation in a separate `.md` file, put tests in a lot of different files, while nvFuser still treats `DynamicType` as a whole.

Because `DynamicType` heavily uses advanced C++ features, we should have a higher standard for `DynamicType` by writing more thorough tests, and run the tests on multiple different platforms (gcc9, gcc10, gcc11, gcc12, gcc13, clang++, C++17, C++20, C++23, C++26). Currently, I already have something like "this function can not be `constexpr` in C++17, but should automatically become `constexpr` in C++20" in mind, but there is no way for me to validate this, because we are using C++17. By making it a separate library, we can build the tests for `DynamicType` on all modern C++ standards, and test if it works. It also makes me possible to add many CIs to test only `DynamicType` on multiple different platforms. It is not practical to test the entire nvFuser on all these platforms, because building the entire nvFuser on GitHub actions today already takes 3 hours, and I don't want to furture add latency.

Generally, I think being modular is a good thing, and we should split more pieces of nvFuser into separate libraries if it makes sense. I would consider the relationship of `DynamicType` vs nvFuser as c10 vs PyTorch.

This PR should be only code movement, there should be no behavioral change.

### CODE REVIEW

**Comment by zasdfgbnm** [comment]
!build

**Comment by zasdfgbnm** [comment]
!build

File: lib/dynamic_type/CMakeLists.txt [hunk 1]
```diff
@@ -0,0 +1,7 @@
+add_library(dynamic_type INTERFACE)
+target_include_directories(dynamic_type INTERFACE src)
+
+add_executable(test_dynamic_type test/test_dynamic_type.cpp)
```
Comment (zasdfgbnm) [inline lib/dynamic_type/CMakeLists.txt:4]: Will our CI run tests for this file?

**Comment by zasdfgbnm** [comment]
!build

**Comment by csarofeen** [comment]
I'd love to see a readme for dynamic type. It's such an interesting library.

File: lib/dynamic_type/CMakeLists.txt (hunk 1 above)
Comment (csarofeen) [reply lib/dynamic_type/CMakeLists.txt:4]: CC @xwang233

**Review by csarofeen — APPROVED** [review]
LGTM

**Pull request merged**